StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift

```diff
@@ -239,20 +241,42 @@ class WebSocketClient_Tests: StressTestCase {
         AssertAsync.staysTrue(engine.connect_calledCount == 0)
     }
     
+    // MARK: - Ping Controller
+    
     func test_pingIsSentPeriodically() {
         // Simulate connection
         test_connectionFlow()
-        
-        let pingInterval = WebSocketClient.pingTimeInterval
+        XCTAssertEqual(pingController.connectionStateDidChangeCount, 3)
+        XCTAssertEqual(pingController.pongRecievedCount, 1)
         assert(engine.sendPing_calledCount == 0)
         
         // Simulate time longer than pingTimeInterval and assert `engine.sendPing` was called
-        time.run(numberOfSeconds: pingInterval + 1)
+        time.run(numberOfSeconds: WebSocketPingController.pingTimeInterval)
+        engine.simulatePong()
         XCTAssertEqual(engine.sendPing_calledCount, 1)
+        XCTAssertEqual(pingController.pongRecievedCount, 2)
+    }
+    
+    func test_pongTimeout() {
+        // Simulate connection
+        test_connectionFlow()
+        XCTAssertEqual(engine.connect_calledCount, 1)
+        
+        reconnectionStrategy.reconnectionDelay = 1
+        var timeInterval = WebSocketPingController.pingTimeInterval
         
-        // Simulate time 3x longer than pingTimeInterval and assert 3 more pings
-        time.run(numberOfSeconds: 3 * pingInterval)
-        XCTAssertEqual(engine.sendPing_calledCount, 1 + 3)
+        // Simulate time longer than pingTimeInterval and assert `engine.sendPing` was called
+        assert(engine.sendPing_calledCount == 0)
+        time.run(numberOfSeconds: timeInterval)
+        XCTAssertEqual(engine.sendPing_calledCount, 1)
+        timeInterval += WebSocketPingController.pongTimeoutTimeInterval
+        time.run(numberOfSeconds: timeInterval)
+        XCTAssertEqual(engine.disconnect_calledCount, 1)
+        XCTAssertFalse(webSocketClient.connectionState.isConnected)
+        
+        AssertAsync {
+            Assert.willBeTrue(self.engine.connect_calledCount == 2)
+        }
     }
```

Comment:
None of this is needed, really :) You already test this functionality in the `WebSocketPingController_Tests`. You can remove these two tests and replace them with:
```
func test_webSocketPingController_ connectionStateDidChange_calledWhenConnectionChanges() {}
func test_webSocketPingController_ping_callsEngineWithPing()
func test_webSocketPingController_forceReconnect_disconnectsEngine()
```
All these tests should be very simple, synchronous, without any mention of virtual time. You just need to test is the integration between `WebSocketClinet` and `WebSocketPingController`, not the functionality of `WebSocketPingController`.